racket-cas/racket-cas.rkt

```diff
@@ -1998,6 +2001,183 @@
+       (for/or ([c candidates])
+         (try-subst-integrate u x c)
+         ))
+     (cond
+       [subst-results subst-results]
+       [else
+        (define expanded (expand u))
+        (if
+         (equal? expanded u)
+         #f
+         (integrate expanded x))]
+       )]))
+
+(define (integrate-basic u x)
+  (when debugging? (displayln (list 'integrate-basic u x)))
+  (math-match u
+              [u #:when (free-of u x)       (⊗ u x)]
+              [(⊗ u v) #:when (free-of u x) (⊗ u (integrate-basic v x))]
+              [(⊕ u v)                      (⊕ (integrate-basic u x) (integrate-basic v x))]
+              [x (⊗ 1/2 (Sqr x))]
+              [(Expt x -1) (Ln x)] ; (Ln (Abs x))
+              [(Expt x u) #:when (free-of u x) (⊘ (Expt x (⊕ 1 u)) (⊕ 1 u))]
+              [(Expt u x) #:when (free-of u x) (⊘ (Expt u x) (Ln u))]
+              [(Ln x)    (⊖ (⊗ (Ln x) x) x)]
+              [(Sin x)   (⊖ (Cos x))]
+              [(Cos x)   (Sin x)]
+              [(Expt (⊕ 1 (Expt x 2)) -1) (list 'atan x)]
+              [_                            (raise 'integrate-basic-failed)]
+              ))
+              
+(module+ test
+  (check-equal? (integrate 1 x) x)
+  (check-equal? (integrate x x) (⊗ 1/2 (Sqr x)))
+  (check-equal? (integrate (Expt x 3) x) (⊗ 1/4 (Expt x 4)))
+  ;-------
+  (check-equal? (integrate '(* (cos x) (sin x)) x) '(* -1/2 (expt (cos x) 2)))
+  ; (check-equal? (integrate '(* (sin x) (sin x)) x) #f) ; to be fixed. with cos 2x
+  (check-equal? (integrate (⊗ 2 x (Cos (Sqr x))) x) '(sin (expt x 2)))
+  (check-equal? (integrate '(* 2 x (expt (+ (expt x 2) 4) 5)) x) '(* 1/6 (expt (expt (+ 4 (expt x 2)) 5) 6/5)))
+  (check-equal? (integrate '(* (cos x) (expt 2 (sin x))) x) '(* (expt 2 (sin x)) (expt (ln 2) -1)))
+  (let* [(u (⊕ x 1)) (v (Sqr u))]
+    (check-equal?
+     (integrate (⊘ (⊗ u (Ln (Cos v)) (Sin v)) (Cos v)) x)
+     '(* -1/4 (expt (ln (cos (expt (+ 1 x) 2))) 2))
+  ))
+  (check-equal? (integrate '(* (+ x 1) (+ x 2)) x) '(+ (* -1/2 (expt (+ 2 x) 2)) (* 1/3 (expt (+ 2 x) 3))))
+  (check-equal? (integrate '(* (+ (* 2 x) 1) (cos (+ x (expt x 2)))) x) '(sin (+ x (expt x 2))))
+  (check-equal? (integrate '(* 5 x (cos (expt x 2)) (sin (expt x 2))) x) '(* 5/4 (expt (sin (expt x 2)) 2)))
+  (check-equal? (integrate '(* (+ (cos x) 2) (+ (sin x) 3)) x)
+                '(+ (* 6 x) (* -2 (cos x)) (* -1/2 (expt (cos x) 2)) (* 3 (sin x))))
+  (check-equal? (integrate '(* (sin x) (expt (cos x) -4)) x) '(* 1/3 (expt (cos x) -3)))
+  (check-equal? (integrate '(* (expt (+ (sin x) 4) 3) (cos x)) x) '(* 1/4 (expt (+ 4 (sin x)) 4)))
+  (check-equal? (integrate '(* (+ x 2) (expt (+ (expt x 2) (* 4 x) 2) -1)) x)
+                '(* -1/2 (ln (expt (+ 2 (* 4 x) (expt x 2)) -1))))
+  ; Fix this with fraction simplification / divide gcd.
+  ; (check-equal? (integrate '(* (sin (+ (* a (expt x 2)) (* b (expt x 2)))) x) x) ...)
+  ; definite integral
+  (check-equal? (integrate `u x `a `b) '(+ (* -1 a u) (* b u)))
+  )
```

Comment:
This is already a nice range of expressions that can be integrated.

If you need inspiration, I can recommend looking at: Yacas
https://yacas.readthedocs.io/en/latest/book_of_algorithms/integration.html
The test suite is here:
https://github.com/grzegorzmazur/yacas/blob/master/tests/integrate.yts
And their rule set is here:
https://github.com/grzegorzmazur/yacas/blob/master/scripts/integrate.rep/code.ys